File: 1456-find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance/find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance.java
class Solution {
    public int findTheCity(int n, int[][] edges, int distanceThreshold) {
        int [][] distances = new int[n][n];
        for(int [] dist : distances){
            Arrays.fill(dist, Integer.MAX_VALUE);
        }
        for(int i = 0;i<n;i++){
            distances[i][i] = 0;
        }
        for(int [] edge : edges){
            distances[edge[0]][edge[1]] = edge[2];
            distances[edge[1]][edge[0]] = edge[2];
        }
        for(int via = 0;via < n;via++){
            for(int src = 0;src<n;src++){
                for(int target = 0;target< n;target++){
                    if(distances[src][via] != Integer.MAX_VALUE && distances[via][target] != Integer.MAX_VALUE){
                        distances[src][target] = Math.min(distances[src][target], distances[src][via] + distances[via][target]);
                    }
                }
            }
        }
        int temp = 0;
        int min = Integer.MAX_VALUE;
        int result = 0;
        for(int city = 0;city<n;city++){
            temp = 0;
            for(int anotherCity = 0;anotherCity < n;anotherCity++){
                if(city != anotherCity && distances[city][anotherCity] <= distanceThreshold){
                    temp+=1;
                }
            }
            if(temp <= min){
                min = temp;
                result = city;
            }
        }
        return result;
    }
}

File: 1456-find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance/find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance.py
class Solution:
    def findTheCity(self, n: int, edges: List[List[int]], distanceThreshold: int) -> int:
        adjMatrix = [[float('inf') if i != j else 0 for i in range(n)] for j in range(n)]
        for edge in edges:
            adjMatrix[edge[0]][edge[1]] = edge[2]
            adjMatrix[edge[1]][edge[0]] = edge[2]
        for i in range(n):
            for r in range(n):
                for c in range(n):
                    if adjMatrix[r][i] != float('inf') and adjMatrix[i][c] != float('inf'):
                        adjMatrix[r][c] = min(adjMatrix[r][c], (adjMatrix[r][i] + adjMatrix[i][c]))
        minCities = n
        result = -1
        for i in range(n):
            count = 0
            for j in range(n):
                if i != j and adjMatrix[i][j] <= distanceThreshold:
                    count+=1
            if count <= minCities:
                result = i
                minCities = count
        return result

                    